modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -109,23 +121,48 @@ public Netty4HttpPipeliningHandler(
     public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
         activityTracker.startActivity();
         try {
-            assert msg instanceof FullHttpRequest : "Should have fully aggregated message already but saw [" + msg + "]";
-            final FullHttpRequest fullHttpRequest = (FullHttpRequest) msg;
             final Netty4HttpRequest netty4HttpRequest;
-            if (fullHttpRequest.decoderResult().isFailure()) {
-                final Throwable cause = fullHttpRequest.decoderResult().cause();
-                final Exception nonError;
-                if (cause instanceof Error) {
-                    ExceptionsHelper.maybeDieOnAnotherThread(cause);
-                    nonError = new Exception(cause);
+            if (msg instanceof HttpRequest request) {
+                if (request.decoderResult().isFailure()) {
+                    final Throwable cause = request.decoderResult().cause();
+                    final Exception nonError;
+                    if (cause instanceof Error) {
+                        ExceptionsHelper.maybeDieOnAnotherThread(cause);
+                        nonError = new Exception(cause);
+                    } else {
+                        nonError = (Exception) cause;
+                    }
+                    netty4HttpRequest = new Netty4HttpRequest(readSequence++, (FullHttpRequest) request, nonError);
                 } else {
-                    nonError = (Exception) cause;
+                    if (request instanceof FullHttpRequest fullHttpRequest) {
+                        currentRequestStream = null;
+                        netty4HttpRequest = new Netty4HttpRequest(readSequence++, fullHttpRequest);
+                    } else {
+                        var contentStream = new Netty4HttpRequestBodyStream(ctx.channel());
+                        currentRequestStream = contentStream;
+                        netty4HttpRequest = new Netty4HttpRequest(
+                            readSequence++,
+                            new DefaultFullHttpRequest(
+                                request.protocolVersion(),
+                                request.method(),
+                                request.uri(),
+                                Unpooled.EMPTY_BUFFER,
+                                request.headers(),
+                                EmptyHttpHeaders.INSTANCE
+                            ),
```

Comment:
Are we just using a `FullHttpRequest` because this is a POC, and to minimise the change? `HttpRequest` would seem more appropriate here.

Reply:
> to minimise the change?

Yes, it's one of the non-essential small things that touches many lines of code and bloat PR. Still need to address this, no reason to pass FullHttpRequest here.